ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -90,4 +105,153 @@ updateLoEFragStall k getCandidates =
+      -- headers after the LoE
+      guard $ AF.lastPoint frag0 /= AF.lastPoint frag1
+      -- peer1 offers more than k blocks
+      guard moreThanK
+      -- peer1 definitely has higher density than peer0
+      guard $ lb1 > ub0
+      pure peer0
+
+    loeIntersectionSlot = AF.headSlot loeFrag
+    -- exclusive last slot in the Genesis window
+    endOfGenesisWindow =
+        succWithOrigin loeIntersectionSlot + SlotNo sgen
+      where
+        GenesisWindow sgen =
+          computeGenesisWindow
+            (configLedger cfg)
+            -- TODO: use a forecasted ledger view for the intersection
+            -- slot (tip of LoE frag).
+            (ledgerState immutableLedgerSt)
+
+    dropBeyondGenesisWindow =
+      AF.takeWhileOldest ((< endOfGenesisWindow) . blockSlot)
+
+    competingFrags =
+      Map.map dropBeyondGenesisWindow candidateSuffixes
+
+    -- Length of an interval with inclusive lower bound @a@ and exclusive upper
+    -- bound @b@.
+    intervalLength a b
+      | a <= b    = b - a
+      | otherwise = 0
+
+    SecurityParam k = configSecurityParam cfg
+
+    showBounds :: (AnchoredFragment (Header blk), Bool, Word64, Word64) -> String
+    showBounds (_, more, lower, upper) = show lower ++ "/" ++ show upper ++ "[" ++ (if more then "+" else " ") ++ "]"
+
+{- TODO: convert this scribble into a useful explanatory diagram, illustrating the
+        density calculation below
+
+            |--------|
+
+    frag1: A - B - C - D - ...            <- exact
+            \
+    frag2:     E           claimed tip: E <- exact
+
+-}
+
+-- | Run the Genesis disconnection logic once.
+--
+--   * Maintain and update the LoE.
+--
+--   * Disconnect from peers with inferior density.
+updateLoEFragGenesis ::
+     forall m blk peer.
+     ( IOLike m
+     , Ord peer
+     , Show peer
+     , LedgerSupportsProtocol blk
+     )
+  => TopLevelConfig blk
+  -> Tracer m String
+  -> STM m (Map peer (AnchoredFragment (Header blk)))
+  -> STM m (Map peer (ChainSyncClientHandle m blk))
+  -> UpdateLoEFrag m blk
+updateLoEFragGenesis cfg tracer getCandidates getHandles =
+  UpdateLoEFrag $ \ curChain immutableLedgerSt setLoEFrag -> do
+    (candidateSuffixes, handles, theirTips, loeFrag) <- atomically $ do
+      candidates <- getCandidates
+      handles <- getHandles
+      let
+        (loeFrag, candidateSuffixes) =
+          sharedCandidatePrefix (configSecurityParam cfg) curChain candidates
+      setLoEFrag loeFrag
+      -- Obtaining these before calling setLoEFrag appeared to result in a wrong state,
+      -- but maybe that was a fluke.
+      theirTips <-
+        fmap (Map.mapMaybe id) $ flip Map.traverseWithKey candidateSuffixes $ \peer _ ->
+          cschTheirTip (handles Map.! peer)
+      pure (candidateSuffixes, handles, theirTips, loeFrag)
+
+    let
+      (losingPeers, boundsDesc) =
+        densityDisconnect cfg immutableLedgerSt candidateSuffixes theirTips loeFrag
```

Comment:
The LoE fragment stays outdated until the next block is added to the chain db. It is still to be justified that this doesn't cause any sort of deadlock because the LoE isn't sufficiently advanced.

Reply:
Moved conversation to #984